core/compiler.common.jvm/src/org/jetbrains/kotlin/load/java/AbstractAnnotationTypeQualifierResolver.kt

```diff
@@ -0,0 +1,205 @@
+        val nullabilityQualifier = extractNullability(typeQualifier) { false } ?: return null
+        return JavaDefaultQualifiers(nullabilityQualifier.copy(isForWarningOnly = jsr305State.isWarning), applicability)
+    }
+
+    fun extractAndMergeDefaultQualifiers(
+        oldQualifiers: JavaTypeQualifiersByElementType?, annotations: Iterable<Annotation>
+    ): JavaTypeQualifiersByElementType? {
+        if (javaTypeEnhancementState.disabledDefaultAnnotations) return oldQualifiers
+
+        val defaultQualifiers = annotations.mapNotNull { extractDefaultQualifiers(it) }
+        if (defaultQualifiers.isEmpty()) return oldQualifiers
+
+        val defaultQualifiersByType =
+            oldQualifiers?.defaultQualifiers?.let(::QualifierByApplicabilityType)
+                ?: QualifierByApplicabilityType(AnnotationQualifierApplicabilityType::class.java)
+
+        var wasUpdate = false
+        for (qualifier in defaultQualifiers) {
+            for (applicabilityType in qualifier.qualifierApplicabilityTypes) {
+                defaultQualifiersByType[applicabilityType] = qualifier
+                wasUpdate = true
+            }
+        }
+
+        return if (!wasUpdate) oldQualifiers else JavaTypeQualifiersByElementType(defaultQualifiersByType)
+    }
+
+    private fun knownNullability(annotation: Annotation, forceWarning: Boolean): NullabilityQualifierWithMigrationStatus? {
+        val fqName = annotation.fqName ?: return null
+        val reportLevel = javaTypeEnhancementState.getReportLevelForAnnotation(fqName)
+        if (reportLevel.isIgnore) return null
+        val nullability = when (fqName) {
+            in NULLABLE_ANNOTATIONS -> NullabilityQualifier.NULLABLE
+            in NOT_NULL_ANNOTATIONS -> NullabilityQualifier.NOT_NULL
+            JSPECIFY_NULLABLE -> NullabilityQualifier.NULLABLE
+            JSPECIFY_NULLNESS_UNKNOWN -> NullabilityQualifier.FORCE_FLEXIBILITY
+            JAVAX_NONNULL_ANNOTATION ->
+                when (annotation.enumArguments(onlyValue = false).firstOrNull()) {
+                    "ALWAYS", null -> NullabilityQualifier.NOT_NULL
+                    "MAYBE", "NEVER" -> NullabilityQualifier.NULLABLE
+                    "UNKNOWN" -> NullabilityQualifier.FORCE_FLEXIBILITY
+                    else -> return null
+                }
+            COMPATQUAL_NULLABLE_ANNOTATION -> NullabilityQualifier.NULLABLE
+            COMPATQUAL_NONNULL_ANNOTATION -> NullabilityQualifier.NOT_NULL
+            ANDROIDX_RECENTLY_NON_NULL_ANNOTATION -> NullabilityQualifier.NOT_NULL
+            ANDROIDX_RECENTLY_NULLABLE_ANNOTATION -> NullabilityQualifier.NULLABLE
+            else -> return null
+        }
+        return NullabilityQualifierWithMigrationStatus(nullability, reportLevel.isWarning || forceWarning)
+    }
+
+    @OptIn(ExperimentalStdlibApi::class)
+    private companion object {
+        val JAVA_APPLICABILITY_TYPES = buildMap<String, AnnotationQualifierApplicabilityType> {
```

Comment:
BTW since 1.6 (which is already used for the compiler source code) we have "unrestricted builder inference" due to which the type arguments here could be eliminated.